core/corehttp/corehttp.go

```diff
@@ -75,8 +79,22 @@ func listenAndServe(node *core.IpfsNode, addr ma.Multiaddr, handler http.Handler
 	// if node being closed before server exits, close server
 	case <-node.Closing():
 		log.Infof("server at %s terminating...", addr)
+
+		// make sure keep-alive connections do not keep the server running
+		server.InnerServer.SetKeepAlivesEnabled(false)
+
 		server.Shutdown <- true
-		<-serverExited // now, DO wait until server exit
+
+	outer:
+		for {
+			// wait until server exits
+			select {
+			case <-serverExited:
+				break outer
+			case <-time.After(5 * time.Second):
+				log.Infof("waiting for server at %s to terminate...", addr)
```

Comment:
I was thinking the `log.Infof("server at %s terminating...", addr)` line at the start was taking care of that part, so that the timeout logging was more of a "still waiting" kind.